tests/utils.py

```diff
@@ -154,43 +147,114 @@ class TestServerProcess():
-    # of the subprocess. In the mode "r+"" stands for reading and writing
+    # of the subprocess. In the mode "r+" stands for reading and writing
     # and "t" stands for text mode.
     self.__temp_log_file = tempfile.TemporaryFile(mode='r+t')
 
     self.server = server
-    self.port = port or random.randint(30000, 45000)
     self.__logger = log
+    try:
+      self._start_server(timeout, extra_cmd_args, popen_cwd)
+      wait_for_server('localhost', self.server, self.port, timeout)
+    except Exception as e:
+      # Clean the resources and log the server errors if any exists.
+      self.clean()
+      raise e
+
+
+
+  def _start_server(self, timeout, extra_cmd_args, popen_cwd):
+    """Start the server subprocess. Uses a retry mechanism
+    if the bind fails."""
+
+    success = False
+    retries = 0
+    elapsed = 0
+    start = time.time()
+    while not success and elapsed < timeout:
+      retries += 1
+      self._start_process(extra_cmd_args,  popen_cwd)
+
+      # loop until bind succeeds, server exits or we timeout
+      while elapsed < timeout:
+        if not self.is_process_running():
+          break
+        elif self._is_port_found_in_log():
+          # If the port is in the log, then the bind was successful.
+          success = True
+          break
+
+        time.sleep(0.01)
+        elapsed = time.time() - start
+
+      if not success:
+        # If the server has not started for whatever reason
+        self.__logger.info('Failed to start ' + self.server + '! Retrying!')
+        self._kill_server_process()
+        self.__temp_log_file.truncate(0)
+
+    if not success:
+      raise TimeoutError('Failure during ' + self.server + ' startup! ' \
+        + 'Made ' + str(retries) + ' retries with random ports!')
+
+    self.__logger.info(self.server + ' serving at ' + str(self.port))
+
+
+
+  def _start_process(self, extra_cmd_args, popen_cwd):
+    """Starts the process running the server."""
 
     # The "-u" option forces stdin, stdout and stderr to be unbuffered.
-    command = ['python', '-u', server, str(self.port)] + extra_cmd_args
+    command = ['python', '-u', self.server] + extra_cmd_args
 
     # We are reusing one server subprocess in multiple unit tests, but we are
     # collecting the logs per test.
     self.__server_process = subprocess.Popen(command,
         stdout=self.__temp_log_file, stderr=subprocess.STDOUT, cwd=popen_cwd)
 
-    self.__logger.info('Server process with process id ' \
-        + str(self.__server_process.pid) + " serving on port " \
-        + str(self.port) + ' started.')
 
-    if timeout > 0:
-      try:
-        wait_for_server('localhost', self.server, self.port, timeout)
-      except Exception as e:
-        # Make sure that errors from the server side will be logged.
-        self.flush_log()
-        raise e
+
+  def _is_port_found_in_log(self):
+    """Checks if the port number is sent from the server subprocess."""
```

Comment:
nit: this also sets self.port -- so I guess the name could imply it's not just a check. I don't have great suggestions.